src/repr/src/row/iter.rs

```diff
@@ -0,0 +1,190 @@
+    /// will not change the value returned from this method.
+    fn count(&self) -> usize;
+
+    /// Maps the returned [`RowRef`]s from this [`RowIterator`].
+    fn map<T, F>(self, f: F) -> MappedRowIterator<Self, F>
+    where
+        Self: Sized,
+        F: FnMut(&RowRef) -> T,
+    {
+        MappedRowIterator {
+            inner: self,
+            func: f,
+        }
+    }
+}
+
+impl<I: RowIterator + ?Sized> RowIterator for Box<I> {
+    fn next(&mut self) -> Option<&RowRef> {
+        (**self).next()
+    }
+
+    fn peek(&mut self) -> Option<&RowRef> {
+        (**self).peek()
+    }
+
+    fn count(&self) -> usize {
+        (**self).count()
+    }
+}
+
+impl<I: RowIterator + ?Sized> RowIterator for &mut I {
+    fn next(&mut self) -> Option<&RowRef> {
+        (**self).next()
+    }
+
+    fn peek(&mut self) -> Option<&RowRef> {
+        (**self).peek()
+    }
+
+    fn count(&self) -> usize {
+        (**self).count()
+    }
+}
+
+#[derive(Debug)]
+pub struct MappedRowIterator<I: RowIterator, F> {
+    inner: I,
+    func: F,
+}
+
+impl<T, F, I: RowIterator> Iterator for MappedRowIterator<I, F>
+where
+    F: FnMut(&RowRef) -> T,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let row_ref = self.inner.next()?;
+        Some((self.func)(row_ref))
+    }
+}
+
+/// Convert a type into a [`RowIterator`].
+pub trait IntoRowIterator {
+    type Iter: RowIterator;
+    fn into_row_iter(self) -> Self::Iter;
+}
+
+impl<T: RowIterator> IntoRowIterator for T {
+    type Iter = Self;
+    fn into_row_iter(self) -> Self::Iter {
+        self
+    }
+}
+
+/// A [`RowIterator`] for a single [`Row`].
+#[derive(Debug)]
+pub struct SingleRowIter {
+    row: Row,
+    finished: bool,
+}
+
+impl RowIterator for SingleRowIter {
+    fn next(&mut self) -> Option<&RowRef> {
+        if self.finished {
+            None
+        } else {
+            self.finished = true;
+            Some(self.row.as_ref())
+        }
+    }
+
+    fn peek(&mut self) -> Option<&RowRef> {
+        if self.finished {
+            None
+        } else {
+            Some(self.row.as_ref())
+        }
+    }
+
+    fn count(&self) -> usize {
+        1
+    }
+}
+
+impl IntoRowIterator for Row {
+    type Iter = SingleRowIter;
+
+    fn into_row_iter(self) -> Self::Iter {
+        SingleRowIter {
+            row: self,
+            finished: false,
+        }
+    }
+}
+
+/// A [`RowIterator`] for a [`Vec`] of [`Row`]s.
+#[derive(Debug, Clone)]
+pub struct VecRowIter {
+    rows: Vec<Row>,
+    index: usize,
+}
+
+impl RowIterator for VecRowIter {
+    fn next(&mut self) -> Option<&RowRef> {
+        let row = self.rows.get(self.index).map(|r| r.as_ref())?;
+        self.index = self.index.saturating_add(1);
+
+        Some(row)
+    }
+
+    fn peek(&mut self) -> Option<&RowRef> {
+        self.rows.get(self.index).map(|r| r.as_ref())
+    }
+
+    fn count(&self) -> usize {
+        self.rows.len()
+    }
```

Comment:
Not covered by any test?

Reply:
makes sense, a `VecRowIter` is rarely used